File: scripts/algorithms/P/Partition Array for Maximum Sum/Partition Array for Maximum Sum.cpp
// Runtime: 392 ms (Top 5.27%) | Memory: 19.5 MB (Top 5.07%)
class Solution {
public:

    int get_max(vector<int>&arr , int s , int e ){
        int mx_val = *max_element(arr.begin() + s , arr.begin() + e + 1);
        return (e-s + 1 )* mx_val ;
    }

    int solve( vector<int>&arr , int idx , int k , vector<vector<int>>&dp){

        if(idx >= arr.size() ) return 0;

        if(dp[idx][k] != -1 )
            return dp[idx][k];

        int ans = 0;

        for(int i = 0 ; i<k ; ++i ){
            if((idx + i ) > arr.size() - 1)
                break;

            int val = get_max(arr , idx , idx + i ) + solve( arr , idx + i + 1 , k , dp );
            ans = max( ans , val );
        }
        return dp[idx][k] = ans ;

    }

    int maxSumAfterPartitioning(vector<int>& arr, int k) {
        vector<vector<int>>dp(arr.size()+1 , vector<int>( k + 1 , -1 ));

        return solve( arr , 0 , k , dp );
    }
};

File: scripts/algorithms/P/Partition Array for Maximum Sum/Partition Array for Maximum Sum..rs
// Runtime: 3 ms (Top 90.0%) | Memory: 2.14 MB (Top 20.0%)

use std::cmp;

impl Solution {
    pub fn max_sum_after_partitioning(arr: Vec<i32>, k: i32) -> i32 {
        let mut k = k as usize;
        let n: usize = arr.len();
        let mut dp: Vec<i32> = vec![0; n + 1];
        dp[1] = arr[0];
        for i in 2..n+1 as usize {
            let mut max_so_far: i32 = arr[i - 1];
            for j in 1..k+1 as usize {
                if i < j {
                    break;
                }
                max_so_far = cmp::max(max_so_far, arr[i - j]);
                dp[i] = cmp::max(dp[i], dp[i - j] + (j as i32) * max_so_far);
            }
        }
        dp[n]
    }
}